src/modules/validator.js

```diff
@@ -53,8 +85,61 @@ class Validator {
         }
     }
 
-    async validateSignature(reviewRecord, expectedPoPRPublicKey = null) {
-        return true; // TODO: implementation
+    async validatePoPRSignaturesAndKeys(popr, expectedPoPRPublicKey = null) {
+        this.chluIpfs.logger.debug('Validating PoPR Signatures and keys');
+        if (!popr.hash) {
+            popr = await this.chluIpfs.reviewRecords.hashPoPR(popr);
+        }
+        const hash = popr.hash;
+        const vmMultihash = this.keyLocationToKeyMultihash(popr.key_location);
+        const isExpectedKey = expectedPoPRPublicKey === null || expectedPoPRPublicKey === vmMultihash;
+        if (!isExpectedKey) {
+            throw new Error('Expected PoPR to be signed by ' + expectedPoPRPublicKey + ' but found ' + vmMultihash);
+        }
+        const mSignature = popr.marketplace_signature;
+        const vSignature = popr.vendor_signature;
+        const vMultihash = this.keyLocationToKeyMultihash(popr.vendor_key_location);
+        const vmSignature = popr.signature;
+        const marketplaceUrl = popr.marketplace_url;
+        const mMultihash = await this.fetchMarketplaceKey(marketplaceUrl);
+        const c = this.chluIpfs.crypto;
+        const validations = await Promise.all([
+            c.verifyMultihash(vMultihash, vmMultihash, vSignature),
+            c.verifyMultihash(mMultihash, vmMultihash, mSignature),
+            c.verifyMultihash(vmMultihash, hash, vmSignature)
+        ]);
+        // false if any validation is false
+        const valid = validations.reduce((acc, v) => acc && v);
+        if (valid) {
+            // Emit events about keys discovered
+            this.chluIpfs.events.emit('vendor pubkey', vMultihash);
```

Comment:
@fazo96 ties back to using promises and not events comment I had earlier. Also ref the event naming convention comment.